### PR TITLE
Nina stiefel canonical

### PR DESCRIPTION
**Pull request by ninamiolane**
Implementing the canonical metric instead of the euclidean metric on stiefel manifolds

### CODE REVIEW

File: geomstats/stiefel.py [hunk 1]
```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
...
+        self.embedding_metric = EuclideanMetric(n*p)
+        self.n = n
+        self.p = p
+
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point):
+        """
+        Canonical inner product on the tangent space at base_point,
+        which is different from the inner product induced by the embedding.
+
+        Formula from:
```
Comment (johmathe) [inline geomstats/stiefel.py]: Nice!

File: geomstats/stiefel.py [hunk 2]
```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
...
+        inner_prod = gs.trace(gs.matmul(aux, tangent_vec_b), axis1=1, axis2=2)
+
+        inner_prod = gs.to_ndarray(inner_prod, to_ndim=2, axis=1)
+        return inner_prod
+
+    def exp(self, tangent_vec, base_point):
+        """
+        Riemannian exponential of a tangent vector wrt to a base point.
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=3)
```
Comment (johmathe) [inline geomstats/stiefel.py]: note for later: we should really be able to refactor all these functions somehow, with a decorator or something similar.

not in this PR, but let's discuss in person

File: geomstats/stiefel.py [hunk 3]
```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
...
+
+        return exp
+
+    def log(self, point, base_point, max_iter=100, tol=1e-6):
+        """
+        Riemannian logarithm of a point wrt a base point.
+
+        Based on:
+        Zimmermann, Ralf
+        "A Matrix-Algebraic Algorithm for the Riemannian Logarithm
```
Comment (johmathe) [inline geomstats/stiefel.py]: can we start adding links to the papers? would be great.

File: geomstats/stiefel.py [hunk 4]
```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
...
+
+        matrix_rd = gs.matmul(matrix_r, gs.transpose(matrix_d, axes=(0, 2, 1)))
+        matrix_v[:, :, p:2*p] = gs.matmul(matrix_v[:, :, p:2*p], matrix_rd)
+        matrix_v = gs.concatenate(
+            [gs.concatenate([matrix_m, matrix_n], axis=1),
+             matrix_v[:, :, p:2*p]],
+            axis=2)
+
+        for k in range(max_iter):
+
```
Comment (johmathe) [inline geomstats/stiefel.py]: leftover space

File: geomstats/stiefel.py [hunk 5]
```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
...
+        matrix_v = gs.zeros((
+            matrix_w.shape[0],
+            max(matrix_w.shape[1], matrix_w.shape[2]),
+            max(matrix_w.shape[1], matrix_w.shape[2])
+            ))
+
+        for i, w in enumerate(matrix_w):
+            matrix_v[i], _ = gs.linalg.qr(w, mode="complete")
+
+        # Procrustes preprocessing
```
Comment (johmathe) [inline geomstats/stiefel.py]: can we split this in a dedicated method?

File: geomstats/stiefel.py [hunk 6]
```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
...
+
+        if n_base_points == 1:
+            base_point = gs.tile(base_point, (n_points, 1, 1))
+        if n_points == 1:
+            point = gs.tile(point, (n_base_points, 1, 1))
+        n_logs = gs.maximum(n_base_points, n_points)
+
+        matrix_m = gs.matmul(gs.transpose(base_point, (0, 2, 1)), point)
+
+        # QR of normal component of a point
```
Comment (johmathe) [inline geomstats/stiefel.py]: maybe split this in a dedicated method and run some unit tests around it?

File: geomstats/stiefel.py [hunk 7]
```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
...
+                or n_tangent_vecs == 1
+                or n_base_points == 1)
+
+        if n_base_points == 1:
+            base_point = gs.tile(base_point, (n_tangent_vecs, 1, 1))
+        if n_tangent_vecs == 1:
+            tangent_vec = gs.tile(tangent_vec, (n_base_points, 1, 1))
+        n_retractions = gs.maximum(n_base_points, n_tangent_vecs)
+        # Q, R = gs.linalg.qr(base_point + tangent_vec)
+        # TODO: remove cycle, when qr will be vectorized
```
Comment (johmathe) [inline geomstats/stiefel.py]: TODO(username)

File: geomstats/stiefel.py [hunk 8]
```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
...
+
+        for k in range(n_liftings):
+            columns_list = []
+            matrix_m_k = matrix_m[k]
+
+            # construct r_0
+            columns_list.append(make_column_r(0, matrix_m_k))
+
+            for i in range(1, n):
+
```
Comment (johmathe) [inline geomstats/stiefel.py]: linbe break

File: geomstats/stiefel.py [hunk 9]
```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
...
+        matrix_m = gs.matmul(gs.transpose(base_point, axes=(0, 2, 1)), point)
+
+        for k in range(n_liftings):
+            columns_list = []
+            matrix_m_k = matrix_m[k]
+
+            # construct r_0
+            columns_list.append(make_column_r(0, matrix_m_k))
+
+            for i in range(1, n):
```
Comment (johmathe) [inline geomstats/stiefel.py]: can we explain why we start at 1 here? Isn't there a way to start from 0 and make things a bit cleaner?

File: tests/test_matrices_space.py [hunk 10]
```diff
@@ -71,6 +71,34 @@ def test_make_symmetric_and_is_symmetric_vectorization(self):
         results = self.space.make_symmetric(mats)
         self.assertTrue(gs.all(self.space.is_symmetric(results)))
 
+    def test_inner_product(self):
+        base_point = gs.array([
```
Comment (johmathe) [inline tests/test_matrices_space.py]: can we try with floating point numbers?

File: geomstats/stiefel.py (hunk 2 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Completely agree.

File: geomstats/stiefel.py (hunk 3 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Done.

File: geomstats/stiefel.py (hunk 4 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Deleted.

File: geomstats/stiefel.py (hunk 9 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Done.

File: geomstats/stiefel.py (hunk 8 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Deleted.

File: geomstats/stiefel.py (hunk 7 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Done.

File: geomstats/stiefel.py (hunk 5 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Done.

File: geomstats/stiefel.py (hunk 6 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Done. Added a TODO for unit tests of auxiliary functions.

File: tests/test_matrices_space.py (hunk 10 above)
Comment (ninamiolane) [reply tests/test_matrices_space.py]: Done.

**Pull request merged**